### ISSUE TITLE
/master/Work%20Areas/Address-Derivation.md

### ISSUE THREAD

**Issue by EAWF** [EAWF/Bitcoin-Merchants-Toolbox#4]
Using the term Xpub to indicate xpub, ypub, and zpub extended public keys, there is an additional field located at the end of the Xpub which is a 8 byte Base58CheckSum, which for our purposes, CAN be ignored.
HOWEVER, it should be mentioned that it is there nonetheless and becomes important in the case that the Xpub is going to be transferred electronically to someone that is not in direct contact with the wallet.
In other words, if I export an XPub from my Mycelium wallet and send it electronically to you to work with, you MUST:
1. Pub = Bin2Hex( Base58_decode( Xpub ))
2. Payload = Left78bytes(Pub)
3. ChkSum = Right4Bytes(Pub)
4. Check = Left4Bytes( SHA256( Hex2Bin( SHA256( Hex2Bin( Payload) ) ) )
5. If  Check !== ChkSum then FAIL else OK

**Comment by Septem151** [comment]
There is definitely a checksum, but it is removed and checked automatically in most if not all standard implementations of bitcoin's [Base58Check encoding](https://en.bitcoin.it/wiki/Base58Check_encoding). If the checksum does not match, the Base58Check function should throw an exception/error.

Also, the checksum is only 32 bits (see [serialization format](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki#serialization-format). As indicated in the BIP, Base58Check is used to encode (which adds the checksum automatically) and decode (checks the checksum for mismatch) for both extended keys and addresses.

If you send me a serialized xpub. I must:
```
raw_xpub = Base58Check_decode(xpub)
if Checksum Mismatch:
  throw Exception
version = first 4 bytes of raw_xpub
if version is not equal to 0x0488b21e (xpub), 0x049d7cb2 (ypub), or 0x04b24746 (zpub):
  throw Exception
depth = 5th byte of raw_xpub
if depth is not equal to 0x03 (3):
  throw Exception
account_public_key = last 33 bytes of raw_xpub
account_chain_code = bytes 14 through 46 of raw_xpub (32 bytes)
// derive external xpub
// derive multiple addresses, or address index specified
```

**Comment by EAWF** [comment]
Jeez, where were you two years ago when I started beating my head against the wall for figuring out this stuff?  Would have saved me at least a half bitcoin in Aleve for my headaches.

OK, the Base58 functions I am using for PHP don't include base58check, just encode and decode, so that's why I mentioned it. 

Good to mention that base58check handles the checksum's automatically maybe?
